resources/augeas.go

```diff
@@ -0,0 +1,249 @@
+	if err != nil {
+		return err
+	}
+	defer obj.recWatcher.Close()
+
+	// notify engine that we're running
+	if err := obj.Running(processChan); err != nil {
+		return err // bubble up a NACK...
+	}
+
+	var send = false // send event?
+	var exit *error
+
+	for {
+		if obj.debug {
+			log.Printf("%s[%s]: Watching: %s", obj.Kind(), obj.GetName(), obj.File) // attempting to watch...
+		}
+
+		select {
+		case event, ok := <-obj.recWatcher.Events():
+			if !ok { // channel shutdown
+				return nil
+			}
+			if err := event.Error; err != nil {
+				return errwrap.Wrapf(err, "Unknown %s[%s] watcher error", obj.Kind(), obj.GetName())
+			}
+			if obj.debug { // don't access event.Body if event.Error isn't nil
+				log.Printf("%s[%s]: Event(%s): %v", obj.Kind(), obj.GetName(), event.Body.Name, event.Body.Op)
+			}
+			send = true
+			obj.StateOK(false) // dirty
+
+		case event := <-obj.Events():
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit // exit
+			}
+			//obj.StateOK(false) // dirty // these events don't invalidate state
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			send = false
+			obj.Event(processChan)
+		}
+	}
+}
+
+// CheckApply method for Augeas resource.
+func (obj *AugeasRes) CheckApply(apply bool) (checkOK bool, err error) {
+	log.Printf("%s[%s]: CheckApply: %s", obj.Kind(), obj.GetName(), obj.File) // attempting to watch...
+	opts := augeas.None
+	if obj.Lens != "" {
+		log.Printf("%s[%s]: NOLOAD NOWAY: %s", obj.Kind(), obj.GetName(), obj.File) // attempting to watch...
+		opts = augeas.NoModlAutoload                                                // if the lens is specified, we can speed up augeas by not loading everything.
+	}
+
+	// Initiate augeas
+	ag, err := augeas.New("/", "", opts)
+	if err != nil {
+		return false, errwrap.Wrapf(err, "Error while initializing augeas!")
+	}
+
+	checkOK = true
+
+	if obj.Lens != "" {
+		// If the lens is set, load the lens for the file we want to edit.
+		if err = ag.Set("/augeas/load/Xmgmt/lens", obj.Lens); err != nil {
+			return false, errwrap.Wrapf(err, "augeas: Error while initializing lens!")
+		}
+		if err = ag.Set("/augeas/load/Xmgmt/incl", obj.File); err != nil {
+			return false, errwrap.Wrapf(err, "augeas: Error while initializing incl!")
+		}
+		if err = ag.Load(); err != nil {
+			return false, errwrap.Wrapf(err, "augeas: Error while loading!")
+		}
+	}
+
+	for _, set := range obj.Sets {
+		// sets:
+		path, value := set[0], set[1]
+		fullpath := fmt.Sprintf("/files/%v/%v", obj.File, path)
+		if getValue, agerr := ag.Get(fullpath); agerr != nil || value != getValue {
+			// note: augeas.Get throws an error if the path does not exist
+			checkOK = false
+			// state is not okay, no work done, exit, but without error
+			if apply {
+				if seterr := ag.Set(fullpath, value); seterr != nil {
+					return false, errwrap.Wrapf(agerr, "Error while setting augeas value!")
+				}
+			}
+		}
+	}
+
+	if !checkOK && apply {
```

Comment:
I'd recommend you stop using the named return values to set and test and instead just flow through the function normally... Doing so will allow you to inline some of the above code and so it won't be as nested. Eg: https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow (also applies to returning early)